File: app/androidApp/app/src/main/java/com/lynxapp/DevelopmentSettings.kt
package com.lynxapp

import android.content.Context
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import java.io.File

/**
 * Debug-only runtime overrides shared by every Lynx page in the Android host.
 *
 * The implementation lives in main because the bundle repository is shared by
 * both build types. Every read is nevertheless guarded by BuildConfig.DEBUG,
 * so a release build cannot observe values left behind by a debug install.
 */
internal object DevelopmentSettings {
    data class BundleServer(
        val bundleId: String,
        val server: String,
    )

    data class Snapshot(
        val bundleServers: List<BundleServer>,
    )

    fun snapshot(context: Context): Snapshot {
        if (!BuildConfig.DEBUG) return Snapshot(emptyList())
        val preferences = context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
        val mappings = runCatching {
            parseMappings(preferences.getString(BUNDLE_SERVERS, "").orEmpty())
        }.getOrDefault(emptyList())
        return Snapshot(mappings)
    }

    fun developmentUrl(context: Context, bundleId: String): String? {
        if (!BuildConfig.DEBUG || !BUNDLE_ID.matches(bundleId)) return null
        val mapping = snapshot(context).bundleServers.firstOrNull { it.bundleId == bundleId }
        return mapping?.let { resolveBundleUrl(bundleId, it.server) }
    }

    /**
     * Agent/automation-driven overrides pushed over adb, one
     * `bundle-id=server-url` per line in the same format the DEV panel
     * serializes. `/data/local/tmp` is traversable (but not listable) by
     * apps and adb-pushed files are world-readable, so a fixed file name
     * works without any permission; on builds where SELinux still denies
     * the read, or on any parse problem, the file is simply ignored — it
     * can never break startup. There is deliberately no cache: `adb push`
     * takes effect on the next page open without a reinstall.
     */
    fun deviceFileDevelopmentUrl(bundleId: String): String? {
        if (!BuildConfig.DEBUG || !BUNDLE_ID.matches(bundleId)) return null
        val mapping = deviceFileMappings().firstOrNull { it.bundleId == bundleId }
        return mapping?.let { resolveBundleUrl(bundleId, it.server) }
    }

    private fun deviceFileMappings(): List<BundleServer> =
        runCatching {
            val file = File(DEVICE_MAPPINGS_FILE)
            if (!file.isFile) return@runCatching emptyList()
            parseMappings(file.readText())
        }.getOrDefault(emptyList())

    /** Returns normalized values suitable for showing again in the list UI. */
    fun save(context: Context, bundleServers: List<BundleServer>): Snapshot {
        check(BuildConfig.DEBUG) { "Development settings are unavailable in release builds" }
        val normalizedMappings = normalizeMappings(bundleServers)
        val serializedMappings = normalizedMappings.joinToString("\n") { mapping ->
            "${mapping.bundleId}=${mapping.server}"
        }
        context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
            .edit()
            .putString(BUNDLE_SERVERS, serializedMappings)
            .apply()
        return Snapshot(normalizedMappings)
    }

    fun validatedBundleServer(bundleId: String, server: String): BundleServer =
        normalizeMapping(BundleServer(bundleId, server), "Bundle server")

    fun loadedBundleIds(context: Context): List<String> {
        if (!BuildConfig.DEBUG) return emptyList()
        return context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
            .getStringSet(LOADED_BUNDLES, emptySet())
            .orEmpty()
            .filter(BUNDLE_ID::matches)
            .sorted()
    }

    @Synchronized
    fun recordLoadedBundle(context: Context, bundleId: String) {
        if (!BuildConfig.DEBUG || !BUNDLE_ID.matches(bundleId)) return
        val preferences = context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
        val loaded = preferences.getStringSet(LOADED_BUNDLES, emptySet())
            .orEmpty()
            .toMutableSet()
        if (loaded.add(bundleId)) {
            preferences.edit().putStringSet(LOADED_BUNDLES, loaded).apply()
        }
    }

    fun clear(context: Context) {
        if (!BuildConfig.DEBUG) return
        context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
            .edit()
            .remove(BUNDLE_SERVERS)
            .apply()
    }

    /** Reads the previous line-based format so existing Debug installs migrate in place. */
    private fun parseMappings(value: String): List<BundleServer> {
        val mappings = mutableListOf<BundleServer>()
        val seen = mutableSetOf<String>()
        value.lineSequence().forEachIndexed { index, originalLine ->
            val line = originalLine.trim()
            if (line.isEmpty() || line.startsWith("#")) return@forEachIndexed
            val separator = line.indexOf('=')
            require(separator > 0 && separator < line.lastIndex) {
                "Line ${index + 1} must use bundle-id=server-url"
            }
            val bundleId = line.substring(0, separator).trim()
            val server = line.substring(separator + 1).trim()
            require(BUNDLE_ID.matches(bundleId)) {
                "Line ${index + 1} has an invalid bundle ID: $bundleId"
            }
            require(seen.add(bundleId)) {
                "Line ${index + 1} repeats bundle ID: $bundleId"
            }
            mappings += BundleServer(
                bundleId = bundleId,
                server = requireHttpUrl(server, "Line ${index + 1}"),
            )
        }
        return mappings
    }

    private fun normalizeMappings(value: List<BundleServer>): List<BundleServer> {
        val seen = mutableSetOf<String>()
        return value.mapIndexed { index, mapping ->
            val normalized = normalizeMapping(mapping, "Entry ${index + 1}")
            require(seen.add(normalized.bundleId)) {
                "Entry ${index + 1} repeats bundle ID: ${normalized.bundleId}"
            }
            normalized
        }
    }

    private fun normalizeMapping(mapping: BundleServer, label: String): BundleServer {
        val bundleId = mapping.bundleId.trim()
        require(BUNDLE_ID.matches(bundleId)) {
            "$label has an invalid bundle ID: $bundleId"
        }
        return BundleServer(
            bundleId = bundleId,
            server = requireHttpUrl(mapping.server.trim(), label),
        )
    }

    private fun requireHttpUrl(value: String, label: String): String {
        val url = value.toHttpUrlOrNull()
        require(url != null && (url.scheme == "http" || url.scheme == "https")) {
            "$label must be a valid http:// or https:// URL"
        }
        return url.toString()
    }

    private fun resolveBundleUrl(bundleId: String, server: String): String {
        val url = checkNotNull(server.toHttpUrlOrNull())
        if (url.encodedPath.endsWith(".lynx.bundle")) return url.toString()
        return url.newBuilder()
            .addPathSegment("$bundleId.lynx.bundle")
            .build()
            .toString()
    }

    private const val PREFERENCES = "lynx.debug.development-settings"
    private const val DEVICE_MAPPINGS_FILE = "/data/local/tmp/lynx_dev_bundles.txt"
    private const val BUNDLE_SERVERS = "bundle-servers"
    private const val LOADED_BUNDLES = "loaded-bundles"
    private val BUNDLE_ID = Regex("^[a-z0-9][a-z0-9-]*$")
}

File: app/iosApp/iosApp/DevelopmentSettings.swift
import Foundation

struct DevelopmentSettingsSnapshot {
  let bundleServers: [BundleServer]
}

struct BundleServer: Equatable {
  let bundleID: String
  let server: String
}

/// Device-local Debug overrides shared by every iOS Lynx page.
enum DevelopmentSettings {
  static var snapshot: DevelopmentSettingsSnapshot {
    #if DEBUG
    let serialized = UserDefaults.standard.string(
      forKey: Keys.bundleServers
    ) ?? ""
    return DevelopmentSettingsSnapshot(
      bundleServers: (try? parseMappings(serialized)) ?? []
    )
    #else
    return DevelopmentSettingsSnapshot(bundleServers: [])
    #endif
  }

  static func developmentURL(for bundleID: String) -> URL? {
    #if DEBUG
    guard isValidBundleID(bundleID),
          let mapping = snapshot.bundleServers.first(where: {
            $0.bundleID == bundleID
          })
    else {
      return nil
    }
    return resolveBundleURL(bundleID: bundleID, server: mapping.server)
    #else
    return nil
    #endif
  }

  @discardableResult
  static func save(
    bundleServers: [BundleServer]
  ) throws -> DevelopmentSettingsSnapshot {
    #if DEBUG
    let mappings = try normalizeMappings(bundleServers)
    let normalizedMappings = mappings
      .map { "\($0.bundleID)=\($0.server)" }
      .joined(separator: "\n")
    UserDefaults.standard.set(normalizedMappings, forKey: Keys.bundleServers)
    return DevelopmentSettingsSnapshot(bundleServers: mappings)
    #else
    throw SettingsError.unavailable
    #endif
  }

  static func validatedBundleServer(
    bundleID: String,
    server: String
  ) throws -> BundleServer {
    try normalizeMapping(
      BundleServer(bundleID: bundleID, server: server),
      label: "Bundle server"
    )
  }

  static var loadedBundleIDs: [String] {
    #if DEBUG
    return (UserDefaults.standard.stringArray(forKey: Keys.loadedBundles) ?? [])
      .filter(isValidBundleID)
      .sorted()
    #else
    return []
    #endif
  }

  static func recordLoadedBundle(_ bundleID: String) {
    #if DEBUG
    guard isValidBundleID(bundleID) else { return }
    loadedBundlesLock.lock()
    defer { loadedBundlesLock.unlock() }
    var loaded = Set(
      UserDefaults.standard.stringArray(forKey: Keys.loadedBundles) ?? []
    )
    guard loaded.insert(bundleID).inserted else { return }
    UserDefaults.standard.set(loaded.sorted(), forKey: Keys.loadedBundles)
    #endif
  }

  static func clear() {
    #if DEBUG
    UserDefaults.standard.removeObject(forKey: Keys.bundleServers)
    #endif
  }

  /// Reads the previous line-based format so existing Debug installs migrate in place.
  private static func parseMappings(_ value: String) throws -> [BundleServer] {
    var mappings: [BundleServer] = []
    var seen = Set<String>()
    for (offset, originalLine) in value.components(separatedBy: .newlines)
      .enumerated() {
      let line = originalLine.trimmingCharacters(in: .whitespaces)
      guard !line.isEmpty, !line.hasPrefix("#") else { continue }
      guard let separator = line.firstIndex(of: "="),
            separator != line.startIndex,
            separator != line.index(before: line.endIndex)
      else {
        throw SettingsError.invalid(
          "Line \(offset + 1) must use bundle-id=server-url"
        )
      }
      let bundleID = String(line[..<separator])
        .trimmingCharacters(in: .whitespaces)
      let server = String(line[line.index(after: separator)...])
        .trimmingCharacters(in: .whitespaces)
      guard isValidBundleID(bundleID) else {
        throw SettingsError.invalid(
          "Line \(offset + 1) has an invalid bundle ID: \(bundleID)"
        )
      }
      guard seen.insert(bundleID).inserted else {
        throw SettingsError.invalid(
          "Line \(offset + 1) repeats bundle ID: \(bundleID)"
        )
      }
      let normalizedServer = try validHTTPURL(
        server,
        label: "Line \(offset + 1)"
      ).absoluteString
      mappings.append(BundleServer(
        bundleID: bundleID,
        server: normalizedServer
      ))
    }
    return mappings
  }

  private static func normalizeMappings(
    _ mappings: [BundleServer]
  ) throws -> [BundleServer] {
    var seen = Set<String>()
    return try mappings.enumerated().map { offset, mapping in
      let normalized = try normalizeMapping(
        mapping,
        label: "Entry \(offset + 1)"
      )
      guard seen.insert(normalized.bundleID).inserted else {
        throw SettingsError.invalid(
          "Entry \(offset + 1) repeats bundle ID: \(normalized.bundleID)"
        )
      }
      return normalized
    }
  }

  private static func normalizeMapping(
    _ mapping: BundleServer,
    label: String
  ) throws -> BundleServer {
    let bundleID = mapping.bundleID.trimmingCharacters(
      in: .whitespacesAndNewlines
    )
    guard isValidBundleID(bundleID) else {
      throw SettingsError.invalid(
        "\(label) has an invalid bundle ID: \(bundleID)"
      )
    }
    let server = mapping.server.trimmingCharacters(
      in: .whitespacesAndNewlines
    )
    return BundleServer(
      bundleID: bundleID,
      server: try validHTTPURL(server, label: label).absoluteString
    )
  }

  private static func validHTTPURL(_ value: String, label: String) throws -> URL {
    guard let components = URLComponents(string: value),
          let scheme = components.scheme?.lowercased(),
          scheme == "http" || scheme == "https",
          components.host != nil,
          let url = components.url
    else {
      throw SettingsError.invalid(
        "\(label) must be a valid http:// or https:// URL"
      )
    }
    return url
  }

  private static func resolveBundleURL(
    bundleID: String,
    server: String
  ) -> URL? {
    guard let url = URL(string: server) else { return nil }
    if url.path.hasSuffix(".lynx.bundle") { return url }
    guard var components = URLComponents(
      url: url,
      resolvingAgainstBaseURL: false
    ) else {
      return nil
    }
    var path = components.percentEncodedPath
    if !path.hasSuffix("/") { path += "/" }
    components.percentEncodedPath = path + "\(bundleID).lynx.bundle"
    return components.url
  }

  private static func isValidBundleID(_ value: String) -> Bool {
    value.range(
      of: "^[a-z0-9][a-z0-9-]*$",
      options: .regularExpression
    ) != nil
  }

  private enum Keys {
    static let bundleServers = "lynx.debug.bundle-servers"
    static let loadedBundles = "lynx.debug.loaded-bundles"
  }

  private static let loadedBundlesLock = NSLock()

  private enum SettingsError: LocalizedError {
    case invalid(String)
    case unavailable

    var errorDescription: String? {
      switch self {
      case let .invalid(message):
        return message
      case .unavailable:
        return "Development settings are unavailable in release builds."
      }
    }
  }
}
